### PR TITLE
Adding GCM driver for the GCM sink.

### PR DESCRIPTION
**Pull request by vmarmol**
Driver performs the authentication and interaction with the GCM	API. The
sink will perform the actual creation and formatting of metrics.

### CODE REVIEW

**Comment by vmarmol** [comment]
Sink is not complete, but this is the first big part of it.

**Comment by vmarmol** [comment]
Ping @vishh @rjnagal I have a few followups to this one ready :)

**Comment by rjnagal** [comment]
LGTM.

Minor questions for my understanding. Merging now.

File: sinks/gcm/auth.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+func getToken() (authToken, error) {
+	rawToken, err := metadata.Get("instance/service-accounts/default/token")
+	if err != nil {
+		return authToken{}, err
+	}
+
+	var token authToken
+	err = json.Unmarshal([]byte(rawToken), &token)
+	if err != nil {
+		return authToken{}, err
```
Comment (vishh) [inline sinks/gcm/auth.go]: nit: Wrap `err` around a more descriptive error message.

File: sinks/gcm/gcm_driver.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+func (self *gcmDriver) refreshToken() error {
+	if time.Now().After(self.tokenExpiration) {
+		token, err := getToken()
+		if err != nil {
+			return nil
+		}
+
+		// Expire the token a bit early.
+		if token.ExpiresIn > 2 {
```
Comment (vishh) [inline sinks/gcm/gcm_driver.go]: nit: Add units here. As discussed offline, we should hopefully not have housekeeping intervals grater than `2 minutes`.

File: sinks/gcm/sink.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		return fmt.Errorf("requesting unrecognized type to be stored in GCM")
+	}
+
+	// TODO(vmarmol): Implement.
+
+	return nil
+}
+
+func (self *GcmSink) GetConfig() string {
+	desc := "Sink type: GCM\n"
```
Comment (vishh) [inline sinks/gcm/sink.go]: May be a TODO to export more debug information?

**Comment by vishh** [comment]
LGTM. We will have to add some tests eventually since this will be production code.

**Pull request merged**